### PR TITLE
Introduce new XCM instruction: `InitiateAssetsTransfer` for complex asset transfers

### PR DESCRIPTION
**Pull request by acatangiu**
This is a continuation of https://github.com/polkadot-fellows/xcm-format/pull/63 following the migration of XCM RFCs to this repo.

# Summary

This RFC proposes a new instruction that provides a way to initiate asset transfers which transfer multiple types (teleports, local-reserve, destination-reserve) of assets, on remote chains using XCM alone.

The currently existing instructions are too opinionated and force each XCM asset transfer to a single transfer type (teleport, local-reserve, destination-reserve). This results in inability to combine different types of transfers in single transfer which results in overall poor UX when trying to move assets across chains.

# Example usage

```rust
    // pay remote fees with ROCs
    let remote_fees = Some(
        AssetTransferFilter::ReserveWithdraw(
            AssetFilter::Wild(AllOf { id: rocs_id.into(), fun: WildFungibility::Fungible })
        )
    );
    // XCM to be executed locally
    let xcm = Xcm::<penpal_runtime::RuntimeCall>(vec![
        // Withdraw both ROCs and PENs from origin account
        WithdrawAsset(assets.clone().into()),
        // Execute the transfers while paying remote fees with ROCs
        InitiateAssetsTransfer {
            dest: local_asset_hub,
            assets: vec![
                // ROCs are reserve-withdrawn on AHR
                ReserveWithdraw(rocs.into()),
                // PENs are teleported to AHR
                Teleport(pens.into()),
            ],
            remote_fees,
            remote_xcm: xcm_on_ahr,
        },
    ]);
```

I also have a [proof-of-concept implementation](https://github.com/acatangiu/polkadot-sdk/blob/1738f8eebd33d87c9c09c7a6bcd7c42fc1a8aa9e/polkadot/xcm/xcm-executor/src/lib.rs#L717C4-L831C6) and a [proof-of-concept asset transfer test](https://github.com/acatangiu/polkadot-sdk/blob/1738f8eebd33d87c9c09c7a6bcd7c42fc1a8aa9e/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/asset_transfers.rs#L484C1-L489C60), transferring 2 different types of assets across 3 chains.

### CODE REVIEW

File: text/0100-xcm-multi-type-asset-transfer.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+///   program, followed by a `BuyExecution(fee)`, then rest of transfers follow.
+///   This guarantees `remote_xcm` will successfully pass a `AllowTopLevelPaidExecutionFrom` barrier.
+/// - `remote_xcm`: Custom instructions that will be executed on the `dest` chain. Note that
+///   these instructions will be executed after a `ClearOrigin` so their origin will be `None`.
+///
+/// Safety: No concerns.
+///
+/// Kind: *Command*.
+///
+/// Errors:
```
Comment (franciscoaguirre) [inline text/0100-xcm-multi-type-asset-transfer.md:107]: None?

File: text/0100-xcm-multi-type-asset-transfer.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+    // XCM to be executed at dest (Westend Asset Hub)
+    let xcm_on_dest =
+        Xcm(vec![DepositAsset { assets: Wild(All), beneficiary: beneficiary.clone() }]);
+
+    // XCM to be executed at Rococo Asset Hub
+    let context = PenpalUniversalLocation::get();
+    let reanchored_assets = assets.clone().reanchored(&local_asset_hub, &context).unwrap();
+    let reanchored_dest = destination.clone().reanchored(&local_asset_hub, &context).unwrap();
+    let reanchored_rocs_id = rocs_id.clone().reanchored(&local_asset_hub, &context).unwrap();
```
Comment (franciscoaguirre) [inline text/0100-xcm-multi-type-asset-transfer.md:173]: `destination` and `rocs_id` are not defined in this example

**Review by franciscoaguirre — APPROVED** [review]

**Comment by xlc** [comment]
not sure if I missed something but what about the order of the asset deposit?
e.g. transfer from Acala to AH, of 50 DOT and 10 USDT. it will result one UMP and one HRMP and on AH, it will receive one DMP and one HRMP. how exactly is this case handled?

**Comment by acatangiu** [comment]
> not sure if I missed something but what about the order of the asset deposit? e.g. transfer from Acala to AH, of 50 DOT and 10 USDT. it will result one UMP and one HRMP at Acala. On AH, it will receive one DMP and one HRMP. how exactly is this case handled?

This new instruction is point-to-point (same as the rest). It can be used by higher layers that build cross chain xcm programs to move assets across multiple hops by calling it at every hop along the way. But the required/used path is outside the scope of concerns of the XCVM. It only executes this instr which explicitly specifies the dest (next hop).

Transferring assets that require different paths is still not supported within same XCM because of the async nature of cross chain messages..

I will include some of this information in the RFC for clarity.

File: text/0100-xcm-multi-type-asset-transfer.md (hunk 1 above)
Comment (acatangiu) [reply text/0100-xcm-multi-type-asset-transfer.md]: the error type is not enforced by the RFC, removed

File: text/0100-xcm-multi-type-asset-transfer.md (hunk 2 above)
Comment (acatangiu) [reply text/0100-xcm-multi-type-asset-transfer.md:176]: I added definitions for `destination` and `rocs_id` for clarity, but please note this is meant as a conceptual example. To actually compile it, it needs much more boiler-plate code..

File: text/0100-xcm-multi-type-asset-transfer.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+Bridge asset transfers greatly benefit from this change by allowing building XCM programs to transfer multiple
+assets across multiple hops in a single pseudo-atomic action.  
+For example, allows single XCM program execution to transfer multiple assets from `ParaK` on Kusama, through Kusama
+Asset Hub, over the bridge through Polkadot Asset Hub with final destination `ParaP` on Polkadot.
+
+With current XCM, we are limited to doing multiple independent transfers for each individual hop in order to
+move both "interesting" assets, but also "supporting" assets (used to pay fees).
+
+**Note:** Transferring assets that require different paths (chains along the way) is _not supported within same XCM_ because of the async nature of cross chain messages. This new instruction, however, enables initiating transfers for multiple assets that take the same path even if they require different transfer types along that path.
```
Comment (xlc) [inline text/0100-xcm-multi-type-asset-transfer.md:48]: should we define the behaviour when not supported parameters are passed? I want to make sure there is no undefined behaviour.
e.g. if the implementation should check and reject it, or if this is implementation-specific behaviour (i.e. out of scope of RFC but should be documented by the XCM executor)

**Comment by franciscoaguirre** [comment]
> not sure if I missed something but what about the order of the asset deposit? e.g. transfer from Acala to AH, of 50 DOT and 10 USDT. it will result one UMP and one HRMP at Acala. On AH, it will receive one DMP and one HRMP. how exactly is this case handled?

If AH is used as a reserve then this would just be one HRMP and it'd be possible, no?

File: text/0100-xcm-multi-type-asset-transfer.md (hunk 3 above)
Comment (franciscoaguirre) [reply text/0100-xcm-multi-type-asset-transfer.md:48]: The actual path messages take is out of scope for the XCVM, it only knows how to send them. The executor should document how it handles this

Comment (acatangiu) [reply text/0100-xcm-multi-type-asset-transfer.md]: I've added more details and clearly specified that this instruction is used point-to-point, and has to be composed to achieve multi-hop transfers.

For each hop, its parameters unequivocally specify what should happen for that specific hop/leg, so there should be no room for undefined behavior.

Any chain-specific checks or validations of the parameters are implementation-specific behavior.

**Review by vgeddes — APPROVED** [review]
The design looks good. I've confirmed in https://hackmd.io/@vgeddes/rkeSGIAYR that `InitiateAssetsTransfer` unlocks new potential for Snowbridge, mainly in improving UX.

File: text/0100-xcm-multi-type-asset-transfer.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+These filters identify the assets to be transferred as well as the transfer type to be used for transferring
+them.
+It shall handle the local side of the transfer, then forward an onward XCM to `dest` for handling
+the remote side of the transfer.
+
+It should do so using same mechanisms as existing `DepositReserveAsset`, `InitiateReserveWithdraw`, `InitiateTeleport`
+instructions but practically combining all required XCM instructions to be remotely executed into a _single_
+remote XCM program to be sent over to `dest`.
+
+Furthermore, through `remote_fees: Option<AssetTransferFilter>`, it shall allow specifying a single asset to be used
```
Comment (vgeddes) [inline text/0100-xcm-multi-type-asset-transfer.md:147]: Just an observation at the implementation level, when `dest` is a bridged location, for example, Ethereum; this instruction does initiate a multi-hop transfer. There is a hidden XCM fee charged by BridgeHub for committing `remote_xcm` for delivery to `dest`, and consequently the AH executor (SovereignPaidRemoteExporter) needs to estimate and take fees to cover that, with a degree of overestimation as xcm fees are variable).

Given that it likely that AH and BH will be collapsed into a single chain (Plaza) at some point in the future (thanks to CoreTime), I think its ok not to complicate the interface of `InitiateAssetsTransfer` with this complexity. 👍

Comment (acatangiu) [reply text/0100-xcm-multi-type-asset-transfer.md:147]: The last commit is exactly trying to clarify that the “dest” on the instruction is not the final destination, just the “next hop” from the point of view of assets movements (e.g. through a reserve chain). It is the destination for the current step of the transfer. Not the final destination.

----

> when dest is a bridged location like Ethereum; this instruction does end up initiating a multi-hop transfer that is somewhat hidden from the builder.

yes but it is “multi-hop” from purely a transport point of view.

from the actual XCM program point of view it is still “directly” AH->Ethereum. The XCM “origin” as seen by Ethereum is AH and the xcm program logic operates as if message was passed directly from AH.

**Comment by acatangiu** [comment]
/rfc process 0xfdf10d88746da6af4245b60405acbd49fbee059d109185b97c7b5fd8541b17c4

**Pull request merged**